src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,163 @@
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> firstNode;
+    private Node<T> lastNode;
+    private int size;
+
+    private static class Node<E> {
+        private E item;
+        private Node<E> next;
+        private Node<E> prev;
+
+        public Node(Node<E> prev, E item, Node<E> next) {
+            this.item = item;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            firstNode = new Node<T>(null, value, null);
+            lastNode = firstNode;
+        } else {
+            lastNode.next = new Node<>(lastNode, value, null);
+            lastNode = lastNode.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, firstNode);
+            firstNode.prev = newNode;
+            firstNode = newNode;
+            size++;
+        } else {
+            Node<T> afterNode = getNode(index);
+            Node<T> beforeNode = afterNode.prev;
+            Node<T> newNode = new Node<T>(beforeNode, value, afterNode);
+            afterNode.prev = newNode;
+            if (beforeNode != null) {
+                beforeNode.next = newNode;
+            } else {
+                firstNode = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldValue = getNode(index).item;
+        getNode(index).item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> removed = getNode(index);
+        T value = removed.item;
+        unlink(removed);
+        return value;
+
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> node = getNode(object);
+        if (node == null) {
+            return false;
+        }
+        unlink(node);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < 0) {
+            throw new IndexOutOfBoundsException("Can't get element by negative index");
+        } else if (index >= size) {
+            throw new IndexOutOfBoundsException("No such element exists");
```

Comment:
You can use your method "checkIndex"

Reply:
Oh. Forgot this moment

Reply:
I think we can put 
` if (index >= size) {
            throw new IndexOutOfBoundsException("No such element exists");` 
 inside of  `checkIndex(index)`.
 If we need to differentiate `>` and `>=` cases, we can call e.g. `checkIndex(index+1)`